proparse/src/main/java/org/prorefactor/proparse/antlr4/PreproEval.java

```diff
@@ -0,0 +1,612 @@
+    if (a.length() == 0 || b.length() == 0)
+      return 0;
+    int startIndex = 0;
+    if (z != null)
+      startIndex = getInt(z);
+    String source = a.toLowerCase();
+    String target = b.toLowerCase();
+    // Progress counts from one, returns zero if not found.
+    // (Java counts from zero, returns -1 if not found.)
+    return source.indexOf(target, startIndex - 1) + 1;
+  }
+
+  static Integer integer(Object o) {
+    if (o == null)
+      return null;
+    if (o instanceof Number)
+      return Math.round(((Number) o).floatValue());
+    if (o instanceof String)
+      return Math.round(getNumber((String) o).floatValue());
+    if (o instanceof Boolean)
+      return (Boolean) o ? 1 : 0;
+    throw new ProEvalException("Error converting to INTEGER.");
+  }
+
+  static String keyword(Object o) {
+    String s = getString(o);
+    int ttype = -2;
+    ttype = NodeTypes.testLiteralsTable(s, ttype);
+    if (ttype > 0 && NodeTypes.isReserved(ttype))
+      return NodeTypes.getFullText(s);
+    return null;
+  }
+
+  static String keywordall(Object o) {
+    String s = getString(o);
+    int ttype = -2;
+    ttype = NodeTypes.testLiteralsTable(s, ttype);
+    if (ttype > 0 && NodeTypes.isKeywordType(ttype))
+      return NodeTypes.getFullText(s);
+    else {
+      // KEYWORD-ALL returns a value even for method and attribute
+      // names, but Proparse doesn't track all those. So, we
+      // never return an unknown here, we always return the uppercased
+      // text of whatever was passed us.
+      return s.toUpperCase();
+    }
+  }
+
+  static String lefttrim(Object a, Object b) {
+    if (b != null) {
+      String t = getString(b);
+      return StringFuncs.ltrim(getString(a), t);
+    }
+    return StringFuncs.ltrim(getString(a));
+  }
+
+  static Integer lookup(Object x, Object y, Object z) {
+    if (x == null || y == null)
+      return null;
+    String a = getString(x);
+    String b = getString(y);
+    if (a.length() == 0 && b.length() == 0)
+      return 1;
+    a = a.toLowerCase();
+    b = b.toLowerCase();
+    String delim;
+    if (z == null)
+      delim = ",";
+    else
+      delim = ((String) z).toLowerCase();
+    Pattern regex = Pattern.compile(delim, Pattern.LITERAL);
+    List<String> expr = Arrays.asList(regex.split(a));
+    List<String> list = Arrays.asList(regex.split(b));
+    if (expr.size() == 1)
+      return list.indexOf(a) + 1;
+    // From the docs:
+    // If expression contains a delimiter, LOOKUP returns the beginning
+    // of a series of entries in list.
+    // For example, LOOKUP("a,b,c","x,a,b,c") returns a 2.
+    int exprSize = expr.size();
+    int listSize = list.size();
+    for (int index = 0; index < listSize; ++index) {
+      int end = index + exprSize;
+      if (end >= listSize)
+        return 0;
+      if (list.subList(index, end).equals(expr))
+        return index + 1;
+    }
+    return 0;
+  }
+
+  static Boolean matches(Object y, Object z) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "matches" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)